src/pg/sql/cdb_utils.sql

```diff
@@ -299,7 +299,7 @@ BEGIN
 
     RETURN (best_result)[2:array_upper(best_result, 1)];
 END;
-$$ language plpgsql IMMUTABLE;
```

Comment:
Are you sure this is `VOLATILE`? the code seems to at least try to provide reproducibility (`setseed` above). Any side effects maybe?

Reply:
Technically both setseed and random are volatile

```
analysis_api_test_db=# \df+ setseed
                                                                                   List of functions
   Schema   |  Name   | Result data type | Argument data types |  Type  | Volatility |  Parallel  |  Owner   | Security | Access privileges | Language | Source code |   Description   
------------+---------+------------------+---------------------+--------+------------+------------+----------+----------+-------------------+----------+-------------+-----------------
 pg_catalog | setseed | void             | double precision    | normal | volatile   | restricted | postgres | invoker  |                   | internal | setseed     | set random seed
(1 row)

analysis_api_test_db=# \df+ random 
                                                                                 List of functions
   Schema   |  Name  | Result data type | Argument data types |  Type  | Volatility |  Parallel  |  Owner   | Security | Access privileges | Language | Source code | Description  
------------+--------+------------------+---------------------+--------+------------+------------+----------+----------+-------------------+----------+-------------+--------------
 pg_catalog | random | double precision |                     | normal | volatile   | restricted | postgres | invoker  |                   | internal | drandom     | random value
```
The issue here is that modifying the seed could be considered as part of the output as that is not cacheable. For performance reasons we could consider setting it as STABLE as done in the quota/redis calls.

Reply:
hm, you're right, a call to `setseed` provokes a side effect (and global state change, namely the random number generator) that prevents it from being parallel safe.